### PR TITLE
feat!: Remove deprecated ProviderNotReady event

### PR DESCRIPTION
**Pull request by maxnrp**
### Intent
Bring the Kotlin SDK in-line with the OpenFeature specification.

- Canonical events don't include `PROVIDER_NOT_READY`: [5.1.1.](https://openfeature.dev/specification/sections/events/#requirement-511), [5.3.5.](https://openfeature.dev/specification/sections/events/#requirement-535).

- `PROVIDER_ERROR` should use structured details (`message` and `code`): [5.1.4.](https://openfeature.dev/specification/sections/events/#requirement-514), [5.1.5.](https://openfeature.dev/specification/sections/events/#requirement-515).

### Motivation
`PROVIDER_NOT_READY` is not a spec event — NOT_READY is a client/provider status. Keeping `OpenFeatureProviderEvents.ProviderNotReady` invited non-portable provider implementations.
`ProviderError.error` was redundant with `EventDetails` and duplicated how errors are represented compared to the spec’s “provider event details” model.

`fromMessageAndErrorCode` previously dropped the custom message for `PROVIDER_NOT_READY`; using `EventDetails` everywhere makes consistent, message-bearing `ProviderNotReadyError` instances possible.

### Changes
- Removed `OpenFeatureProviderEvents.ProviderNotReady` and updated `MultiProvider` so aggregate `NOT_READY` comes from initial/lifecycle state, not from a synthetic event.

- `ProviderError` now has only `eventDetails`;
- `toOpenFeatureStatusError()` simplified accordingly.
- `OpenFeatureError.fromMessageAndErrorCode: PROVIDER_NOT_READY` now constructs `ProviderNotReadyError(errorMessage)` so the message from `EventDetails` is preserved.

### Tests
- **MultiProviderTests**: precedence scenario for `NOT_READY` over Error/Stale rewritten so one child stays NOT_READY without emitting the removed event; aggregate precedence test updated after dropping `ProviderNotReady` from another scenario.

- **EventDetailsTests**, **ProviderEventingTests**, **AutoHealingProvider**: migrated to `ProviderError(EventDetails(...))` (including `PROVIDER_NOT_READY` with explicit messages).

### Breaking changes
- `ProviderNotReady` event has been removed.
- `ProviderError` no longer accepts an `error` value. Instead, it now only accepts an optional `EventDetails` object (`eventDetails: EventDetails?`).

### CODE REVIEW

**Comment by typotter** [comment]
**missing-test-null-errorcode**

nit: No test for `ProviderError(EventDetails(message="test", errorCode=null))`, the branch in `toOpenFeatureStatusError()` where `details` is non-null but `errorCode` is null (early-return at line 45).

File: docs/multiprovider/README.md [hunk 1]
```diff
@@ -60,7 +60,7 @@ Pick the strategy that best matches your failure-policy:
 Children are evaluated in the order provided. Put the most authoritative or fastest provider first. For example, place a small in-memory override provider before a remote provider to reduce latency.
 
 ### Events and status aggregation
-`MultiProvider` listens to child provider events and emits a single, aggregate status via `OpenFeatureAPI.statusFlow`. The highest-precedence status among children wins:
+`MultiProvider` listens to child provider events and emits a single, aggregate status via `OpenFeatureAPI.statusFlow`. Per the OpenFeature specification: a child stays `NOT_READY` until it emits `PROVIDER_READY`, `PROVIDER_ERROR`, or `PROVIDER_STALE` (or only `PROVIDER_CONFIGURATION_CHANGED`, which does not change readiness). The highest-precedence status among children wins:
```
Comment (typotter) [inline docs/multiprovider/README.md:68]: There is no migration note for callers currently using `ProviderNotReady` or `ProviderError.error`. Since both are removed in this PR, it would help to add a brief callout here (or link to the changelog) so readers know what to replace them with.


```

File: kotlin-sdk/src/commonTest/kotlin/dev/openfeature/kotlin/sdk/EventDetailsTests.kt [hunk 2]
```diff
@@ -41,35 +41,41 @@ class EventDetailsTests {
...
 
         val status = evt.toOpenFeatureStatusError()
-        val fatal = assertIs<OpenFeatureStatus.Fatal>(status)
-        val err = assertIs<OpenFeatureError.ProviderFatalError>(fatal.error)
+        val errorStatus = assertIs<OpenFeatureStatus.Error>(status)
+        val err = assertIs<OpenFeatureError.InvalidContextError>(errorStatus.error)
         assertEquals("message", err.message)
     }
 
     @Test
```
Comment (typotter) [inline kotlin-sdk/src/commonTest/kotlin/dev/openfeature/kotlin/sdk/EventDetailsTests.kt:58]: Is it worth it to add a test that verifies the error message fromn OpenFeatureError propagates past `toOpenFeatureStatusError`?


```

File: kotlin-sdk/src/commonTest/kotlin/dev/openfeature/kotlin/sdk/EventDetailsTests.kt [hunk 3]
```diff
@@ -41,35 +41,41 @@ class EventDetailsTests {
...
-            error = OpenFeatureError.InvalidContextError("message")
-        )
+    fun providerErrorMapToUnspecifiedError() {
+        val evt = OpenFeatureProviderEvents.ProviderError()
 
         val status = evt.toOpenFeatureStatusError()
         val fatal = assertIs<OpenFeatureStatus.Error>(status)
-        val err = assertIs<OpenFeatureError.InvalidContextError>(fatal.error)
-        assertEquals("message", err.message)
+        assertIs<OpenFeatureError.GeneralError>(fatal.error)
```
Comment (typotter) [inline kotlin-sdk/src/commonTest/kotlin/dev/openfeature/kotlin/sdk/EventDetailsTests.kt]: consider asserting on the message string as well

File: kotlin-sdk/src/commonMain/kotlin/dev/openfeature/kotlin/sdk/events/OpenFeatureProviderEvents.kt [hunk 4]
```diff
@@ -26,8 +25,7 @@ sealed class OpenFeatureProviderEvents {
      * The provider signaled an error.
```
Comment (typotter) [inline kotlin-sdk/src/commonMain/kotlin/dev/openfeature/kotlin/sdk/events/OpenFeatureProviderEvents.kt:26]: The KDoc on `ProviderError` predates the removal of `error` and doesn't mention `eventDetails`.

File: kotlin-sdk/src/commonMain/kotlin/dev/openfeature/kotlin/sdk/events/OpenFeatureProviderEvents.kt [hunk 5]
```diff
@@ -40,35 +38,20 @@ sealed class OpenFeatureProviderEvents {
...
-        error != null -> { // Deprecated implementation
-            if (error is ProviderFatalError) {
-                OpenFeatureStatus.Fatal(error)
-            } else {
-                OpenFeatureStatus.Error(error)
-            }
-        }
-
-        else -> OpenFeatureStatus.Error(OpenFeatureError.GeneralError("Unspecified error"))
+    val details = eventDetails
```
Comment (typotter) [inline kotlin-sdk/src/commonMain/kotlin/dev/openfeature/kotlin/sdk/events/OpenFeatureProviderEvents.kt]: `val details = eventDetails` is redundant; Kotlin will smart-cast it after the null check. Can remove the alias and use `eventDetails` directly throughout.

File: kotlin-sdk/src/commonTest/kotlin/dev/openfeature/kotlin/sdk/EventDetailsTests.kt (hunk 2 above)
Comment (maxnrp) [reply kotlin-sdk/src/commonTest/kotlin/dev/openfeature/kotlin/sdk/EventDetailsTests.kt:58]: Yes, and also, `fatal` is misleading here.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,12 @@
 # Changelog
 
+## Unreleased
+
+### ⚠ BREAKING CHANGES
+
+* Removed deprecated `ProviderNotReady` event and `ProviderError.error` [#.#.#](https://github.com/open-feature/kotlin-sdk/pull/239).
```
Comment (nicklasl) [inline CHANGELOG.md]: Just FYI: I _think_ release please will overwrite this.

**Review by nicklasl — APPROVED** [review]

File: CHANGELOG.md (hunk 6 above)
Comment (maxnrp) [reply CHANGELOG.md]: I see. I moved the warning to the readme file.

**Comment by typotter** [comment]
### Code review

No issues found. Checked for bugs and CLAUDE.md compliance.

Cross-SDK note: removing `ProviderNotReady` is spec-compliant — the OpenFeature spec defines four provider events (`PROVIDER_READY`, `PROVIDER_ERROR`, `PROVIDER_STALE`, `PROVIDER_CONFIGURATION_CHANGED`) and `NOT_READY` is a status, not an event. All other SDKs (Java, JS, .NET, Swift) confirm the same four-event model with no `NOT_READY` event. Similarly, the `EventDetails(message, errorCode)` error carrier aligns with how every other SDK structures provider error details.

One minor documentation note (not blocking): the breaking-changes bullet in `docs/multiprovider/README.md` says `ProviderError` "now optionally accepts an `EventDetails` object" — `eventDetails` was already present before this PR; only the `error` field is being removed. Worth a quick wording fix before merge to avoid confusing migrating consumers.

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: docs/multiprovider/README.md [hunk 7]
```diff
@@ -70,28 +75,40 @@ Children are evaluated in the order provided. Put the most authoritative or fast
 
 `ProviderConfigurationChanged` is re-emitted as-is. When the aggregate status changes due to a child event, the original triggering event is also emitted.
 
+⚠ Breaking changes:
```
Comment (typotter) [inline docs/multiprovider/README.md]: I think I misdirected you when commenting about the migration path. The README isn't actually a good place for this; it should only contain information about the state of the codebase _now_ as opposed to breaking changes made in the past. This would be fine to be in the PR description. The changelog will point to the PR as a breaking change, so readers can get these instructions from there.

File: docs/multiprovider/README.md [hunk 8]
```diff
@@ -70,28 +75,40 @@ Children are evaluated in the order provided. Put the most authoritative or fast
 
 `ProviderConfigurationChanged` is re-emitted as-is. When the aggregate status changes due to a child event, the original triggering event is also emitted.
 
+⚠ Breaking changes:  
+
+- `ProviderNotReady` event has been removed.
+- `ProviderError.error` no longer accepts an `error` value. Instead, it now optionally accepts an `EventDetails` object (`eventDetails: EventDetails?`).
```
Comment (typotter) [inline docs/multiprovider/README.md]: ProviderError.error always accepted an EventDetails object. only the lack of `error` is new.

```suggestion
- `ProviderError.error` no longer accepts an `error` value. Instead, it now only accepts an optional `EventDetails` object (`eventDetails: EventDetails?`).
```

**Review by typotter — APPROVED** [review]
